### ISSUE TITLE
IPv6 capabilities causes impurity in nix build

### ISSUE THREAD

**Issue by AleXoundOS** [NixOS/nix#5615]
**Describe the bug**

nix build outcome depends on IPv6 build system (linux) capabilities!

In particular, some nixpkgs `python39Packages.sanic` test fails when being built on NixOS machine with:
```nix
{
  boot.kernelParams = [ "ipv6.disable=1" ];
  networking.enableIPv6 = false;
}
```

Whereas, NixOS machine with IPv6 enabled passes the test and thus produces build result without errors.

**Steps To Reproduce**

With flakes: `nix build github:nixos/nixpkgs/64065d76f434457073f5d255a3246658119e08ed#python39Packages.sanic`.
Without flakes: `nix-build -I nixpkgs=https://github.com/NixOS/nixpkgs/archive/64065d76f434457073f5d255a3246658119e08ed.tar.gz '<nixpkgs>' -A python39Packages.sanic`.

**Expected behavior**

nix build must be reproducible and not affected by network capabilities (except fetching sources/cache).

**`nix --version` output**
`nix (Nix) 2.4pre20201205_a5d85d0`

**Additional context**

[ #146682](https://github.com/NixOS/nixpkgs/pull/146682) revealed the issue.

**Comment by regnat** [comment]
From what I can see, disabling IPv6 means setting a couple of sysctls ([here](https://github.com/NixOS/nixpkgs/blob/24528474d2b3370f2f23879a557ae2cc92a5d50b/nixos/modules/tasks/network-interfaces.nix#L1122-L1123))

I guess it would be an easy fix if these can be tweaked per-namespace (and very hard otherwise)

**Comment by AleXoundOS** [comment]
It's unclear, tweaked in which direction? If system has IPv6 disabled can it be enabled in a namespace?

**Comment by regnat** [comment]
> It's unclear, tweaked in which direction? If system has IPv6 disabled can it be enabled in a namespace?

I was thinking of locally disabling it, since as you mention it’s always possible to forcibly disable it globally by tweaking the kernel params. Which obviously wouldn’t make the test from https://github.com/NixOS/nixpkgs/pull/146682 pass but would remove the impurity.

**Comment by AleXoundOS** [comment]
Looks like the easiest solution. But maybe this IPv6 issue is just one of a broader class of impurity that needs to be investigated and solved in a more general way rather than ad-hoc.

**Comment by leo60228** [comment]
FWIW: I don't think `ipv6.disable=1` is any different from the more obvious issues caused by something like `CONFIG_ADVISE_SYSCALLS=n`.

**Comment by leo60228** [comment]
Also see `runCommand "uname" {} "uname -a > $out"`, which feels like fundamentally the same impurity.

**Comment by AleXoundOS** [comment]
Maybe we should refer to some general nix purity policy? Are such impurities going to be solved or where is the borderline (despite moving towards [CAS](https://www.tweag.io/blog/2020-09-10-nix-cas/))?

**Comment by L-as** [comment]
There are a ton of impurities like this, e.g. you can check the kernel version. I don't know whether there's a way of creating a completely deterministic environment with Linux, but I don't think there is.

**Comment by AleXoundOS** [comment]
I mean maybe there is a tracked list of impurities which we're aware of. Maybe some can be potentially avoided, some not. So that disabling IPv6 tests in a package is not an ad-hoc decision but compliant to some policy (despite this exact case is simple though).

**Comment by regnat** [comment]
> I mean maybe there is a tracked list of impurities which we're aware of. Maybe some can be potentially avoided, some not. So that disabling IPv6 tests in a package is not an ad-hoc decision but compliant to some policy (despite this exact case is simple though).

I don’t think so. That would be big because there’s all sort of impurities can appear at the kernel level. https://reproducible-builds.org/ tracks the most frequent ones

(also, https://r13y.com/ watches the reproducibility status of a small subset of nixpkgs, might show some common non-reproducibility issues)

**Comment by L-as** [comment]
I'll throw this idea out since I don't have time to do it currently, but I wish that Nix supported a wasm32-wasi system or similar so that you could have completely reproducible builds. There are some issues wrt. missing tail call support in most WASM JIT compilers, but that's likely going to be fixed relatively soon.

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
